openage/convert/smp.pyx

```diff
@@ -0,0 +1,837 @@
+        cdef Py_ssize_t dpos = first_cmd_offset
+
+        # is the end of the current row reached?
+        cdef bool eor = False
+
+        cdef uint8_t cmd
+        cdef uint8_t nextbyte
+        cdef uint8_t lower_crumb
+        cdef int pixel_count
+
+        # work through commands till end of row.
+        while not eor:
+            if row_data.size() > expected_size:
+                raise Exception(
+                    "Only %d pixels should be drawn in row %d "
+                    "with frame type %d, but we have %d "
+                    "already!" % (
+                        expected_size, rowid,
+                        self.info.frame_type,
+                        row_data.size()
+                    )
+                )
+
+            # fetch drawing instruction
+            cmd = self.get_byte_at(dpos)
+
+            # Last 2 bits store command type
+            lower_crumb = 0b00000011 & cmd
+
+            # opcode: cmd, rowid: rowid
+
+            if lower_crumb == 0b00000011:
+                # eol (end of line) command, this row is finished now.
+                eor = True
+
+                # shadows sometimes need an extra pixel at
+                # the end
+                if row_data.size() < expected_size:
+                    # copy the last drawn pixel
+                    # (still stored in nextbyte)
+                    #
+                    # TODO: confirm that this is the
+                    #       right way to do it
+                    row_data.push_back(pixel(color_shadow,
+                                             nextbyte, 0, 0, 0))
+
+                continue
+
+            elif lower_crumb == 0b00000000:
+                # skip command
+                # draw 'count' transparent pixels
+                # count = (cmd >> 2) + 1
+
+                pixel_count = (cmd >> 2) + 1
+
+                for _ in range(pixel_count):
+                    row_data.push_back(pixel(color_transparent, 0, 0, 0, 0))
+
+            elif lower_crumb == 0b00000001:
+                # color_list command
+                # draw the following 'count' pixels
+                # pixels are stored as rgba 32 bit values
+                # count = (cmd >> 2) + 1
+
+                pixel_count = (cmd >> 2) + 1
+
+                for _ in range(pixel_count):
+
+                    dpos += 1
+                    nextbyte = self.get_byte_at(dpos)
+
+                    row_data.push_back(pixel(color_shadow,
+                                             nextbyte, 0, 0, 0))
+
+            else:
+                raise Exception(
+                    "unknown smp shadow frame drawing command: " +
+                    "%#x in row %d" % (cmd, rowid))
+
+            # process next command
+            dpos += 1
+
+        # end of row reached, return the created pixel array.
+        return
+
+
+cdef class SMPOutlineFrame(SMPFrame):
+    """
+    SMPFrame for the outline graphics.
+    """
```

Comment:
hm, i wonder why they are doing that. in any way, if the outline information is separate from the graphics info, we have to merge them in order to do our alpha-encoding.

Reply:
Probably because they wanted a smaller number of commands (?) Or something to do with their expanded rendering. Merging should work fine, since he outlines does not overlay with any other pixels in the main graphic.